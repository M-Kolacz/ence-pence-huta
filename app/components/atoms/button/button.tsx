import { Slot } from '@radix-ui/react-slot'
import { cva, type VariantProps } from 'class-variance-authority'
import { forwardRef, type ButtonHTMLAttributes } from 'react'

import { cn } from '#app/utils/misc.tsx'

const buttonVariants = cva(
	'inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50',
	{
		variants: {
			variant: {
				primary: 'bg-primary-600 text-neutral-50 hover:bg-primary-600/90',
				secondary:
					'bg-secondary-600 text-neutral-0 hover:bg-secondary-600/90 focus-visible:ring-secondary-600',
				outline:
					'border border-primary-600 bg-neutral-0 text-primary-600 hover:bg-primary-100/30',
				'secondary-outline':
					'border border-secondary-600 bg-neutral-0 text-secondary-600 hover:bg-secondary-100/30 focus-visible:ring-secondary-600',
			},
			size: {
				default: 'h-[40px] px-4 py-3',
			},
		},
		defaultVariants: {
			variant: 'primary',
			size: 'default',
		},
	},
)

export interface ButtonProps
	extends ButtonHTMLAttributes<HTMLButtonElement>,
		VariantProps<typeof buttonVariants> {
	asChild?: boolean
}

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
	({ className, variant, size, asChild = false, ...props }, ref) => {
		const Comp = asChild ? Slot : 'button'
		return (
			<Comp
				className={cn(buttonVariants({ variant, size, className }))}
				ref={ref}
				{...props}
			/>
		)
	},
)
Button.displayName = 'Button'

export { Button, buttonVariants }
