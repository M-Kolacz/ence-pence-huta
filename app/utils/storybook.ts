import { type ComponentType, type ComponentProps } from 'react'
import { type MY_VIEWPORTS, allModes } from '../../.storybook/modes.ts'

export const disableControls = <Component extends ComponentType<any>>(
	controls: (keyof ComponentProps<Component>)[],
) => {
	const disabledOptions = {}

	controls.forEach(control =>
		Object.assign(disabledOptions, {
			[control]: {
				table: {
					disable: true,
				},
			},
		}),
	)

	return disabledOptions
}

export const setViewport = (device: keyof typeof MY_VIEWPORTS) => ({
	viewport: {
		defaultViewport: device,
	},
	chromatic: {
		modes: {
			[device]: allModes[device],
		},
	},
})
