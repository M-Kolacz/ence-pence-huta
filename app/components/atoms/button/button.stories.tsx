import { type Meta, type StoryObj } from '@storybook/react'
import { Icon } from '#app/components/atoms'
import { disableControls, setDesignPreview } from '#app/utils/storybook.ts'
import { variations, sizes } from './button-variants.ts'
import { Button } from './button.tsx'

const meta = {
	title: '⚛️ Atoms/Button',
	component: Button,
	parameters: {
		layout: 'centered',
		...setDesignPreview(
			'https://www.figma.com/file/zx5FkmiDiV7F8Jn5phnpeJ/Ence-Pence-Huta-v2?node-id=20-76&t=SgllQF4sLHtxY4V0-4',
		),
	},
	args: { children: 'Click me', variant: 'default', size: 'default' },
	argTypes: {
		variant: {
			options: variations,
			control: { type: 'radio' },
		},
		size: {
			options: sizes,
			control: { type: 'radio' },
		},
		...disableControls<typeof Button>(['asChild']),
	},
	tags: ['autodocs'],
} satisfies Meta<typeof Button>

export default meta
type Story = StoryObj<typeof meta>

export const Component: Story = {
	args: {
		variant: 'default',
	},
}

export const IconButton: Story = {
	args: {
		variant: 'outline',
		size: 'icon',
		children: <Icon name="baby-shoes" />,
	},
	argTypes: {
		...disableControls<typeof Button>(['size', 'children', 'variant']),
	},
}

export const Variants: Story = {
	render: args => {
		return (
			<div className="flex gap-4">
				{variations.map(variant => (
					<Button key={variant} {...args} variant={variant} />
				))}
			</div>
		)
	},
	argTypes: {
		...disableControls<typeof Button>(['variant']),
	},
}

export const Sizes: Story = {
	render: args => {
		return (
			<div className="flex gap-4">
				{sizes.map(size => (
					<Button
						key={size}
						{...args}
						size={size}
						children={size === 'icon' ? null : args.children}
					/>
				))}
			</div>
		)
	},
	argTypes: {
		...disableControls<typeof Button>(['size']),
	},
}
