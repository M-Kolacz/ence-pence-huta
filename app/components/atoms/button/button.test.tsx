import { faker } from '@faker-js/faker'
import { composeStories } from '@storybook/react'
import { render, screen } from '@testing-library/react'
import { axe } from 'jest-axe'
import { getStoriesTestCases } from '#app/utils/storybook.ts'
import * as ButtonStories from './button.stories'

const { Component: ButtonStory } = composeStories(ButtonStories)

const testCases = getStoriesTestCases(ButtonStories)

describe('Button component', () => {
	test.each(testCases)(
		'snapshot test of %s story',
		async (_storyName, Story) => {
			if (!Story) throw new Error('Missing story to test!')

			const { container } = render(<Story />)
			expect(container).toMatchSnapshot()
		},
	)

	test.each(testCases)('a11y test of %s story', async (_storyName, Story) => {
		if (!Story) throw new Error('Missing story to test!')

		const { container } = render(<Story />)

		const result = await axe(container)

		expect(result).toHaveNoViolations()
	})

	test('should not be clickable', () => {
		const buttonText = faker.lorem.words(2)

		render(<ButtonStory disabled={true}>{buttonText}</ButtonStory>)

		const button = screen.getByRole('button', {
			name: new RegExp(buttonText, 'i'),
		})

		expect(button).toBeDisabled()
	})
})
