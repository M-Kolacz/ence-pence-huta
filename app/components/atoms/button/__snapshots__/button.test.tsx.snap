// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button component > snapshot test of Component story 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-primary-600 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary-600 text-neutral-50 hover:bg-primary-600/90 h-[40px] px-4 py-3"
  >
    Click me
  </button>
</div>
`;

exports[`Button component > snapshot test of IconButton story 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 w-10"
  >
    <svg
      class="inline h-6 w-6 self-center"
    >
      <use
        href="/app/components/atoms/icon/icons/sprite.svg#baby-shoes"
      />
    </svg>
  </button>
</div>
`;

exports[`Button component > snapshot test of Sizes story 1`] = `
<div>
  <div
    class="flex gap-4"
  >
    <button
      class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-9 rounded-md px-3"
    >
      Click me
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
    >
      Click me
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-11 rounded-md px-8"
    >
      Click me
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 w-10"
    />
  </div>
</div>
`;

exports[`Button component > snapshot test of Variants story 1`] = `
<div>
  <div
    class="flex gap-4"
  >
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-primary-600 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary-600 text-neutral-50 hover:bg-primary-600/90 h-[40px] px-4 py-3"
    >
      Click me
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-secondary-600 text-neutral-0 hover:bg-secondary-600/90 focus-visible:ring-secondary-600 h-[40px] px-4 py-3"
    >
      Click me
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-primary-600 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-primary-600 bg-neutral-0 text-primary-600 hover:bg-primary-100/30 h-[40px] px-4 py-3"
    >
      Click me
    </button>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-secondary-600 bg-neutral-0 text-secondary-600 hover:bg-secondary-100/30 focus-visible:ring-secondary-600 h-[40px] px-4 py-3"
    >
      Click me
    </button>
  </div>
</div>
`;
